### ISSUE TITLE
 ObservableRemotePluginServerAdapter.java json protocol backwxard compatibility failed

### ISSUE THREAD

**Issue by angelsephirot** [eclipse-keyple/keyple-distributed-remote-java-lib#15]
if apiLevel is not present  into the client messageDto,  the deserialization automatically set the version to 3
as a consequence  the backward compatibility does not work.

the cclientDistributedApiLevel is set to 3.
By the way the apiLevel must be <0 in the messageDTO to ensure backward compatibility!!.
Is there a reason for both version apiLevel and clientDistributedApiLevel?

    // Creates a remote reader based on the incoming message.
    JsonObject body = JsonUtil.getParser().fromJson(message.getBody(), JsonObject.class);

    // The API level is retrieved from the wrapper, as the body content has been created by the
    // Distributed client layer.
    int clientDistributedApiLevel;
    if (message.getApiLevel() != 0) {
      clientDistributedApiLevel = message.getApiLevel();
    } else if (body.has(JsonProperty.SERVICE_ID.getKey())) {
      clientDistributedApiLevel = 1;
    } else {
      clientDistributedApiLevel = 0;
    }

**Comment by andrei-cristea** [comment]
Hello,

The **Keyple Distributed solution** is made up of two independent components: the **core** which manages the content of the messages exchanged, and the **message wrapper** managed by the lib implementing the Distributed solution.

As a result, there are internal API levels, the **Core API Level** and the **Distributed API Level**.

In **Reader Client Side** usage mode, these two versions are evaluated by the server on the first call as follows: 

```java
    // The API level is retrieved from the wrapper, as the body content has been created by the
    // Distributed client layer.
    int clientDistributedApiLevel;
    if (message.getApiLevel() != 0) {
      clientDistributedApiLevel = message.getApiLevel();
    } else if (body.has(JsonProperty.SERVICE_ID.getKey())) {
      clientDistributedApiLevel = 1;
    } else {
      clientDistributedApiLevel = 0;
    }

    // In this particular case, the API level contained in the body does not reflect the version of
    // the body, but that of the Core client layer.
    int clientCoreApiLevel;
    if (body.has(JsonProperty.CORE_API_LEVEL.getKey())) {
      clientCoreApiLevel = body.get(JsonProperty.CORE_API_LEVEL.getKey()).getAsInt();
    } else {
      clientCoreApiLevel = -1; // Unknown at this step
    }
```

Consequently, if the apiLevel field is not present in the MessageDto, then the server considers that the client is in Distributed API Level 1 or 0, but not 3.

The server remains backwards compatible with clients using older versions.

So I don't understand what type of error you're currently experiencing. We can schedule a meeting to discuss this in person if necessary.

Regards

**Comment by andrei-cristea** [comment]
Fixed by release [2.5.1](https://github.com/eclipse-keyple/keyple-distributed-remote-java-lib/releases/tag/2.5.1)
